src/mount-support.c

```diff
@@ -293,3 +294,55 @@ void setup_slave_mount_namespace()
 		die("can not make make / rslave");
 	}
 }
+
+void setup_bind_mounts(const char *appname)
+{
+	debug("%s: %s", __FUNCTION__, appname);
+
+	FILE *f = NULL;
+	const char *bind_profile_dir = "/var/lib/snapd/bind/profiles/";
+
+	char profile_path[PATH_MAX];
+	int snprintf_rc =
+	    snprintf(profile_path, sizeof(profile_path), "%s/%s.bind",
+		     bind_profile_dir, appname);
+	if (snprintf_rc < 0 || snprintf_rc >= 512) {
+		errno = 0;
+		die("snprintf returned unexpected value");
+	}
+
+	f = fopen(profile_path, "r");
+	// it is ok for the file to not exist
+	if (f == NULL && errno == ENOENT)
+		return;
+	// however any other error is a real error
+	if (f == NULL) {
+		die("cannot open %s", profile_path);
+	}
+
+	struct mntent *m = NULL;
+	while ((m = getmntent(f)) != NULL) {
+		int flags = MS_BIND;
+
+		if (strcmp(m->mnt_type, "") != 0) {
+			die("only bind mounts are supported");
+		}
+		if (hasmntopt(m, "bind") == NULL) {
+			die("need bind mount flag");
+		}
+		if (hasmntopt(m, "ro") != NULL) {
+			flags |= MS_RDONLY;
+		}
+
+		if (mount(m->mnt_fsname, m->mnt_dir, NULL, flags, NULL) != 0) {
```

Comment:
This is where real vulnerabilities could creep in. I need a little more time to think about the source and destination mount points. There are two possible attacks that come to mind.

1. An unprivileged user tricks the running-as-root setup_bind_mounts() to gain privileges on the system by setting up malicious symlinks in the source mount point path.
2. A malicious snap modifies a rw bind mount and drops in a malicious symlink that is then followed by setup_bind_mounts() the next time the snap's bind mounts are set up.

We protect against the 2nd attack in LXC by very carefully chdir()'ing to the destination mount point, verifying inode ownership and not following symlinks along the way.

The 1st attack is a tough one. Hopefully we wouldn't need bind mount a directory owned by a non-root user and could simply enforce that in the code. What are your thoughts on that?